python/lib/core/dmod/core/meta_data.py

```diff
@@ -698,6 +878,67 @@ def factory_init_from_restriction_collections(cls, data_format: DataFormat, **kw
                           continuous_restrictions=None if len(continuous) == 0 else continuous,
                           discrete_restrictions=None if len(discrete) == 0 else discrete)
 
+    @classmethod
+    def merge_domains(cls, d1: 'DataDomain', d2: 'DataDomain') -> 'DataDomain':
+        """
+        Merge the two domains into a new combined domain.
+
+        Parameters
+        ----------
+        d1: DataDomain
+            The first domain.
+        d2: DataDomain
+            The second domain.
+
+        Returns
+        -------
+        DataDomain
+            The new merged domain.
+
+        Raises
+        ------
+        ValueError
+            If the formats or constraints of the two domains do not permit them to be merged
+
+        Notes
+        -----
+        For any two domains ``d1`` and ``d2`` that can successfully be merged, and the derived domain ``d3`` equal to
+        ``merge_domains(d1, d2)``,  then it should always be true that ``substract_domains(d3, d2) == d1``.
+
+        See Also
+        --------
+        subtract_domains
+        """
+        # TODO: (later) consider exceptions to format rule and perhaps other behavior, like for composites
+        if d1.data_format != d2.data_format:
+            raise DmodRuntimeError(f"Can't merge {d2.data_format.name} format domain into one of {d1.data_format.name}")
+
+        # New continuous; taken directly from domain 1 if not present or equal in domain 2; otherwise, by extending
+        new_c_rests = {idx: (d1_rest
+                             if idx not in d2.continuous_restrictions or d1_rest == d2.continuous_restrictions[idx]
+                             else
+                             d1_rest.extend(d2.continuous_restrictions[idx]))
+                       for idx, d1_rest in d1.continuous_restrictions.items()}
```

Comment:
Break into a traditional `foreach` with an inner `if` statement. Only use an `value if blank else other_value` statement if it fits on one line and try to avoid doing nested comprehensions.

Reply:
It is typically a better practice, but (and this involves changes for the next comment too) below is what that ends up looking like.  I feel like the version with the shorter variable names, comprehension, and ternary is actually more readable, even if you have to stop and think for a moment about what it's doing.   That is definitely subjective though, so I'm interested in your opinions @christophertubbs and @aaraney.

```Python
        new_c_rests = dict()
        for idx, base_domain_rest in base_domain.continuous_restrictions.items():
            if (idx not in incorporated_domain.continuous_restrictions or base_domain_rest ==
                    incorporated_domain.continuous_restrictions[idx]):
                new_c_rests[idx] = base_domain_rest
            else:
                new_c_rests[idx] = base_domain_rest.expand(incorporated_domain.continuous_restrictions[idx])
```

Reply:
I'd go with something like:

```python
        new_c_rests = dict()

        for idx, base_domain_rest in base_domain.continuous_restrictions.items():
            index_not_in_continuous_restrictions = idx not in incorporated_domain.continuous_restrictions
            restriction_is_base_domain = base_domain_rest == incorporated_domain.continuous_restrictions.get(idx)

            if index_not_in_continuous_restrictions or restriction_is_base_domain:
                new_c_rests[idx] = base_domain_rest
            else:
                new_c_rests[idx] = base_domain_rest.expand(incorporated_domain.continuous_restrictions[idx])
```

or

```python
        new_c_rests = dict()

        for idx, base_domain_rest in base_domain.continuous_restrictions.items():
            if (idx not in incorporated_domain.continuous_restrictions 
                    or base_domain_rest == incorporated_domain.continuous_restrictions[idx]):
                new_c_rests[idx] = base_domain_rest
            else:
                new_c_rests[idx] = base_domain_rest.expand(incorporated_domain.continuous_restrictions[idx])
```

or 

```python
        new_c_rests = dict()
        for idx, base_domain_rest in base_domain.continuous_restrictions.items():
            use_base_domain = (
                idx not in incorporated_domain.continuous_restrictions
                or base_domain_rest == incorporated_domain.continuous_restrictions[idx]
            )

            if use_base_domain:
                new_c_rests[idx] = base_domain_rest
            else:
                new_c_rests[idx] = base_domain_rest.expand(incorporated_domain.continuous_restrictions[idx])
```

I may not have any idea of what `idx` is or should be and I may not know what `base_domain_rest` is and I may not know why you'd use the base domain rest if the index isn't in the incorporated restrictions or is already in the incorporated restrictions, but I can follow:

1. We have a new dictionary
2. We're going to step through every key value in a map
3. If the key isn't in whatever incorporated domain's continuous restrictions is or the value is the same for its index, just assign the value to the key in the new dictionary
4. Otherwise assign the combination of the base domain and the incorporated domain's continuous restrictions for that index to that index in the new dictionary

It's pretty good for "This happens first, this happens second, this happens third, etc" whereas the original had me bouncing back and forth between the iterator and the ternary to try and track what's going on.

What you provided is fine **except** for the hanging expression:

```python
            if (idx not in incorporated_domain.continuous_restrictions or base_domain_rest ==
                    incorporated_domain.continuous_restrictions[idx]):
```

If you need the new line, break at the `or` operator, not the `==` operator.

Having looked at some examples to make sure I'm on the right track, I think the following may work better and still fulfill the comprehension that you're leaning towards:

```python
new_c_rests = {
    idx: (
        d1_rest
        if idx not in d2.continuous_restrictions or d1_rest == d2.continuous_restrictions[idx]
        else d1_rest.extend(d2.continuous_restrictions[idx])
    )
    for idx, d1_rest in d1.continuous_restrictions.items()
}
```

Same exact code, but each block of logic is nicely separated and lined up. Don't be afraid to take up vertical space put those brackets on new lines, end lines with brackets, same with parenthesis, line up innards, etc.

This bit might be super personal preference, but comprehensions just sort of make an extra level of sense when
they are formatted like sql or linq queries.

Bonus: I'll give you a high five if you annotate like:

```python
        new_c_rests: typing.Dict[MilkyWay, ContinuousRestriction] = dict()

        for idx, base_domain_rest in base_domain.continuous_restrictions.items():  # type: MilkyWay, ContinuousRestriction
```

Then I'll know what `idx` is and I'll have friendly IDE inspections :D

Reply:
I've done some refactoring based on your last suggestion, @christophertubbs, with the adjusted alignment and added type hints (and slightly more indicative variable names).  Let me know what you think.